cpu/native/ng_net/dev_eth_tap.c

```diff
@@ -0,0 +1,299 @@
+ * @author  Kaspar Schleiser <kaspar@schleiser.de>
+ * @}
+ */
+#include <err.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <inttypes.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+
+#ifdef __MACH__
+#define _POSIX_C_SOURCE
+#include <net/if.h>
+#undef _POSIX_C_SOURCE
+#include <ifaddrs.h>
+#include <net/if_dl.h>
+#elif defined(__FreeBSD__)
+#include <sys/socket.h>
+#include <net/if.h>
+#include <ifaddrs.h>
+#include <net/if_dl.h>
+#else
+#include <net/if.h>
+#include <linux/if_tun.h>
+#include <linux/if_ether.h>
+#endif
+
+#include "native_internal.h"
+
+#include "net/dev_eth.h"
+#include "dev_eth_tap.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+/* support one tap interface for now */
+dev_eth_tap_t dev_eth_tap;
+
+/* dev_eth interface */
+static int _init(dev_eth_t *ethdev);
+static int _send(dev_eth_t *ethdev, char* buf, int n);
+static int _recv(dev_eth_t *ethdev, char* buf, int n);
+
+static inline void _get_mac_addr(dev_eth_t *ethdev, uint8_t *dst) {
+    dev_eth_tap_t *dev = (dev_eth_tap_t*)ethdev;
+    memcpy(dst, dev->addr, NG_ETHERNET_ADDR_LEN);
+}
+
+static inline int _get_promiscous(dev_eth_t *ethdev) {
+    dev_eth_tap_t *dev = (dev_eth_tap_t*)ethdev;
+    return dev->promiscous;
+}
+
+static inline int _set_promiscous(dev_eth_t *ethdev, int value) {
+    dev_eth_tap_t *dev = (dev_eth_tap_t*)ethdev;
+    dev->promiscous = value;
+    return value;
+}
+
+static inline void _isr(dev_eth_t *ethdev) {
+    dev_eth_rx_handler(ethdev);
+}
+
+static eth_driver_t eth_driver_tap = {
+    .init = _init,
+    .send = _send,
+    .recv = _recv,
+    .get_mac_addr = _get_mac_addr,
+    .get_promiscous = _get_promiscous,
+    .set_promiscous = _set_promiscous,
+    .isr = _isr,
+};
+
+/* driver implementation */
+static inline bool _is_addr_broadcast(uint8_t *addr)
+{
+    return ((addr[0] == 0xff) && (addr[1] == 0xff) && (addr[2] == 0xff) &&
+            (addr[3] == 0xff) && (addr[4] == 0xff) && (addr[5] == 0xff));
+}
+
+static inline bool _is_addr_multicast(uint8_t *addr)
+{
+    /* source: http://ieee802.org/secmail/pdfocSP2xXA6d.pdf */
+    return (addr[0] & 0x01);
+}
+
+static int _recv(dev_eth_t *dev_eth, char *buf, int len) {
+    dev_eth_tap_t *dev = (dev_eth_tap_t*)dev_eth;
+
+    int nread = real_read(dev->tap_fd, buf, len);
+    DEBUG("ng_tapnet: read %d bytes\n", nread);
+
+    if (nread > 0) {
+        ng_ethernet_hdr_t *hdr = (ng_ethernet_hdr_t *)buf;
+        if (!(dev->promiscous) && !_is_addr_multicast(hdr->dst) &&
+            !_is_addr_broadcast(hdr->dst) &&
+            (memcmp(hdr->dst, dev->addr, NG_ETHERNET_ADDR_LEN) != 0)) {
+            DEBUG("ng_eth_dev: received for %02x:%02x:%02x:%02x:%02x:%02x\n"
+                  "That's not me => Dropped\n",
+                  hdr->dst[0], hdr->dst[1], hdr->dst[2],
+                  hdr->dst[3], hdr->dst[4], hdr->dst[5]);
+            return 0;
```

Comment:
Yes, I think so. At least that worked when I used the sighandler on Linux (I'm writing an alternative tap-like driver right now) but you might need to prevent switching because the polling process could fire immediately afterwards under load.